debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/parser/LogMinerDmlParser.java

```diff
@@ -157,53 +152,32 @@ private LogMinerDmlEntry parseUpdate(String sql, Table table) {
+            List<LogMinerColumnValue> newValues = new ArrayList<>(table.columns().size());
+            index = parseSetClause(sql, index, newValues);
 
             // parse where
-            List<String> oldColumnNames = new ArrayList<>();
-            List<String> oldColumnValues = new ArrayList<>();
-            parseWhereClause(sql, index, oldColumnNames, oldColumnValues);
-
-            final Map<String, String> beforeColumnMap = createMap(oldColumnNames, oldColumnValues);
-            final Map<String, String> afterColumnMap = createMap(newColumnNames, newColumnValues);
-
-            // set before
-            final List<LogMinerColumnValue> oldValues;
-            if (!beforeColumnMap.isEmpty()) {
-                oldValues = new ArrayList<>(table.columns().size());
-                for (int i = 0; i < table.columns().size(); ++i) {
-                    String columnName = table.columns().get(i).name();
-                    String columnValue = beforeColumnMap.get(columnName);
-                    oldValues.add(createColumnValue(columnName, columnValue));
-                }
-            }
-            else {
-                oldValues = Collections.emptyList();
-            }
+            List<LogMinerColumnValue> oldValues = new ArrayList<>(table.columns().size());
+            parseWhereClause(sql, index, oldValues);
 
             // set after
-            List<LogMinerColumnValue> newValues;
-            if (!afterColumnMap.isEmpty()) {
-                newValues = new ArrayList<>(table.columns().size());
+            if (!newValues.isEmpty()) {
                 for (int i = 0; i < table.columns().size(); ++i) {
-                    String columnName = table.columns().get(i).name();
-                    if (afterColumnMap.containsKey(columnName)) {
-                        LogMinerColumnValue value = new LogMinerColumnValueImpl(columnName);
-                        value.setColumnData(afterColumnMap.get(columnName));
-                        newValues.add(value);
-                    }
-                    else {
-                        LogMinerColumnValue value = new LogMinerColumnValueImpl(columnName);
-                        value.setColumnData(beforeColumnMap.get(columnName));
-                        newValues.add(value);
+                    final Column column = table.columns().get(i);
+                    if (newValues.stream().noneMatch(c -> c.getColumnName().equals(column.name()))) {
+                        LogMinerColumnValue value = new LogMinerColumnValueImpl(column.name());
+                        Optional<LogMinerColumnValue> oldValue = oldValues.stream().filter(c -> c.getColumnName().equals(column.name())).findFirst();
+                        oldValue.ifPresent(logMinerColumnValue -> {
+                            value.setColumnData(logMinerColumnValue.getColumnData());
+                            newValues.add(value);
+                        });
                     }
```

Comment:
I'd suggest to generally avoid streams and also `Optional` (= another allocation) on this sensitive code path, rather using the more traditional patterns with explicit looping, early abort if feasible, and null checking. A bit less elegant perhaps, but more appropriate for the case at hand.

Reply:
Could you expand a bit about what is the purpose of this code to begin with? All the name-based access is going to kill us, I'm afraid. Is it ensured that the SET and the WHERE arrays have the same length? If so, we could simply use index-based access for instance.

Reply:
No, this is to ensure that we emit correct state for an `UPDATE` statement.  Lets take the following as an example: 
```
update "DEBEZIUM"."TABLEA" set "COLA" = '25' WHERE "ID" = '1' AND "COLA" = '20';
```
We parse the `SET` clause first and so the new values array only contains `{"COLA"="25"}` where-as the `WHERE` clause gets parsed afterward and contains `{"ID"="1"},{"COLA"="20"}`.  This specific loop looks at the values in the old values from the where-clause and copies a reference to the column name/value into the new values array if that column doesn't exist, so the new values would then contain `{"COLA"="25"},{"ID"="1"}`.